File: skipper-lib/skipper_lib/events/event_receiver.py
import pika
import requests
import json


class EventReceiver(object):
    def __init__(self, username, password, host, port, queue_name, service, service_name, logger):
        self.service_worker = service
        self.service_name = service_name
        self.queue_name = queue_name
        self.logger = logger

        credentials = pika.PlainCredentials(username, password)
        connection = pika.BlockingConnection(pika.ConnectionParameters(host=host,
                                                                       port=port,
                                                                       credentials=credentials))

        channel = connection.channel()
        channel.queue_declare(queue=queue_name)

        channel.basic_qos(prefetch_count=1)
        channel.basic_consume(queue=queue_name, on_message_callback=self.on_request)

        print("Awaiting requests from [x] " + queue_name + " [x]")
        channel.start_consuming()

    def on_request(self, ch, method, props, body):
        service_instance = self.service_worker()

        self.logger_helper(props.correlation_id, self.queue_name, self.service_name, self.logger, 'start', '-')

        try:
            response, task_type = service_instance.call(body)

            ch.basic_publish(exchange='',
                             routing_key=props.reply_to,
                             properties=pika.BasicProperties(correlation_id=props.correlation_id),
                             body=response)
            ch.basic_ack(delivery_tag=method.delivery_tag)

            self.logger_helper(props.correlation_id, self.queue_name, self.service_name, self.logger, 'end', '-')

            print('Processed request:', task_type)
        except Exception as e:
            response = {"error": 'Receiver exception',
                        "queue": self.queue_name,
                        "service_name": self.service_name,
                        "correlation_id": props.correlation_id
                        }
            result = json.dumps(response)

            self.logger_helper(props.correlation_id, self.queue_name, self.service_name, self.logger, 'end', 'Receiver exception')
            print('Receiver exception')

            ch.basic_publish(exchange='',
                             routing_key=props.reply_to,
                             properties=pika.BasicProperties(correlation_id=props.correlation_id),
                             body=result)
            ch.basic_ack(delivery_tag=method.delivery_tag)

    def logger_helper(self, corr_id, queue_name, service_name, logger, task_type, description):
        if self.logger is not None:
            params = {"correlation_id": corr_id,
                      "queue_name": queue_name,
                      "service_name": service_name,
                      "task_type": task_type,
                      "description": description
                      }
            try:
                requests.post(logger, json=params)
            except requests.exceptions.RequestException as e:
                print('Logger service is not available')

File: skipper-lib-js/skipper/events/event_receiver.js
const amqp = require('amqplib/callback_api');
const http = require("http");

class EventReceiver {
    constructor(username, password, host, port, queueName, serviceName) {
        this.connect_url = 'amqp://' + username + ':' + password + '@' + host + ':' + port;
        this.queueName = queueName;
        this.serviceName = serviceName;
    }

    startListener(callback, service, logger, loggerHelper) {
        var queueName = this.queueName;
        var serviceName = this.serviceName;
        amqp.connect(this.connect_url, function (error0, connection) {
            if (error0) {
                throw error0;
            }
            connection.createChannel(function (error1, channel) {
                if (error1) {
                    throw error1;
                }
                var queue = queueName;

                channel.assertQueue(queue, {
                    durable: false
                });
                channel.prefetch(1);
                console.log('[x] Awaiting requests for: ' + serviceName + ' [x]');
                channel.consume(queue, function reply(msg) {
                    var data = msg.content.toString();
                    var r = callback(data, service, logger, queue, serviceName, msg.properties.correlationId, loggerHelper);

                    channel.sendToQueue(msg.properties.replyTo,
                        Buffer.from(r.toString()), {
                        correlationId: msg.properties.correlationId
                    });

                    channel.ack(msg);
                });
            });
        });
    }

    onRequest(data, service, logger, queue, serviceName, correlationId, loggerHelper) {
        loggerHelper(logger, queue, serviceName, correlationId, 'start', '-');

        var result = null;
        try {
            var r = service.call(data);

            loggerHelper(logger, queue, serviceName, correlationId, 'end', '-');
            console.log('Processed request: ' + r[1]);

            result = r[0];
        } catch (error) {
            const response = {
                'error': 'Receiver exception',
                'queue': queue,
                'service_name': serviceName,
                'correlation_id': correlationId
            }
            result = JSON.stringify(response);

            loggerHelper(logger, queue, serviceName, correlationId, 'end', 'Receiver exception');
            console.error("Receiver exception");
        }

        return result;
    }

    loggerHelper(logger, queue, serviceName, correlationId, taskType, description) {
        if (logger !== null) {
            const postOptions = {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
            };

            var requestBody = {
                "correlation_id": correlationId,
                "queue_name": queue,
                "service_name": serviceName,
                "task_type": taskType,
                "description": description

            };

            const req = http.request(
                logger,
                postOptions
            );

            req.on("error", (err) => {
                console.error("Logger service is not available");
            });

            req.write(JSON.stringify(requestBody));
            req.end();
        }
    }
}

module.exports = EventReceiver